### ISSUE TITLE
Security vulnerable in client-side pagination json format

### ISSUE THREAD

**Issue by bashlakov** [wenzhixin/bootstrap-table#1410]
By this article, json format currently used with client-side pagination is vulnerable for some attacks:
http://flask.pocoo.org/docs/0.10/security/#json-security

I think resolution could be to use additional option to change behavior on client-side pagination (use nested array or not).
Additional question is what default value to use: use nesting array by default will provide more security but will request additional actions to enable backward-capability (I think this is more prefered as far as security always must be more prefered).

PS. I can make pull-request for this feature.

What do you think about this?

**Comment by bashlakov** [comment]
In my fork i made this:
https://github.com/bashlakov/bootstrap-table/commit/89f738632ab7c618e4de9b19635956bf832914d8

**Comment by dabros** [comment]
@bashlakov - correct me if wrong, but with client side no json data is sent, just the raw json parsed from the url...

Without side-pagination='server', just using client side, there is no security because there is nothing to secure, no data calls being made to the server and all json data already exposed in unecryted network request and local js data for bootstrap table.

The data is already in that json url (file or server response) and is never affected by anything on the client side.

What security risk are you protecting against - be specific as to a replicable set of steps that a user might do that constitutes a risk...

**Comment by bashlakov** [comment]
If you use such json in admin interface or private pages with https, you probably want to secure this data.

Step-by-step guide to implement this attack:
1) Run backend that genegate json in current format and return them at url http(s)://example.com/api/get_friends.
2) Make page http://hackerpage.com/ with exploit code from link above.
3) By visiting hacker's page, js code on that page will get all top-level objects from json, returned by back-end.

This attack vector aslo described here: http://incompleteness.me/blog/2007/03/05/json-is-not-as-safe-as-people-think-it-is/

**Comment by bashlakov** [comment]
@wenzhixin, could you please review this proposal? I ready to make pr, the question is only default value

**Comment by wenzhixin** [comment]
Hi @bashlakov, I tested the code what article: http://flask.pocoo.org/docs/0.10/security/#json-security said, but the `captured` was empty.
Browser: Chrome 44.0.2403.130 (64-bit).

BTW: Your Step-by-step guide, the url http://hackerpage.com/ return: 
> Welcome to: hackerpage.com
This Web page is parked for FREE, courtesy of GoDaddy.com.

**Comment by wenzhixin** [comment]
Try this:
```js
function Array() { alert("hi"); }
var a = [ 43 ];
```
The constructor of array isn't called.

**Comment by dabros** [comment]
@bashlakov - same points as above

its all good saying 'secure your data with https', but that is absolutely irrelevant once that data is loaded into the javascript $.bootstrapTable object, as https or not a simple F12 and breakpoint shows everything returned from that json request.

The point in security here would be perhaps securing any requests for that url, so that you use ajaxOptions to pass hashes or whatever needed to secure the request and/or same-server-origin type security (session variables and ect)

Nothing to do with json-security that you linked, so far as I can tell, as ALL information returned from your request immediately becomes loaded into the javascript object

There is the point of network requests being intercepted, yes, but in that case you have 'responseHandler' or similar (check doco and search closed&open issues here) that then uses whatever you want to prepare the data before being loaded into the table.


In short - create an actual use-case, with clear step-by-steps that translate into a clear actual security risk, because frankly im not seeing any need for base code changes or any gaping security risk (https or not) that is the responsibility of bootstrapTable.js that you wouldnt handle with 'responseHandler' if needed.

If you can use 'responseHandler' or similar to mimic an easily applicable approach for encryption support (or similar), then maybe we can merge it as an option. But ive already explained that aside from 'man-in-the-middle' (request intercepts) there is no security risk.

**Comment by wenzhixin** [comment]
I am agree with @dabros, if that is the security problem I will merge your pull request, but if not, the option is unnecessary, you can set the global defaults for the table like this example: https://github.com/wenzhixin/bootstrap-table-examples/blob/master/issues/457.html.
For example:
```js
$.extend($.fn.bootstrapTable.defaults, {
    responseHandler: function (res) {
        if (this.sidePagination === 'client') {
            return res.rows;
        }
        return res;
    }
});
```

**Comment by wenzhixin** [comment]
Thanks anyway, close the issue.
